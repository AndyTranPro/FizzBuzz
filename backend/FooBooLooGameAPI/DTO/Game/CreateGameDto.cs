namespace FooBooLooGameAPI.DTO.Game;

public class CreateGameDto
{
    public string Name { get; set; } = string.Empty;
    public string Author { get; set; } = string.Empty;
    public int Min { get; set; }
    public int Max { get; set; }
    public List<GameRuleDto> RuleSet { get; set; }
}
