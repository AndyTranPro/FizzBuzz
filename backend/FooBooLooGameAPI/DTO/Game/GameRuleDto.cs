using FooBooLooGameAPI.Helpers;
using Microsoft.AspNetCore.Mvc;

namespace FooBooLooGameAPI.DTO.Game;

[ModelBinder(BinderType = typeof(GameRuleDtoModelBinder))]
public class GameRuleDto
{
    public int Divisor { get; set; }
    public string Replacement { get; set; }
}