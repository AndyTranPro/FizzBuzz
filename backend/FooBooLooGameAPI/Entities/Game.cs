namespace FooBooLooGameAPI.Entities;

public class Game
{
    public int GameId { get; set; }

    public string Name { get; set; } = string.Empty;

    public string Author { get; set; } = string.Empty;

    public int Min { get; set; }

    public int Max { get; set; }

    public List<GameRule> RuleSet { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}
