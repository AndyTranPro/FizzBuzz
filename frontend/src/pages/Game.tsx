import React, { useState } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import { useAppDispatch } from '../features/store';
import { beginSession } from '../features/sessionSlice';
import { Typography, TextField, Button, Box, Grid } from '@mui/material';
import SportsEsportsIcon from '@mui/icons-material/SportsEsports';
import PersonIcon from '@mui/icons-material/Person';
import TimerIcon from '@mui/icons-material/Timer';
import { keyframes } from '@emotion/react';

// Keyframes for animations
const bounce = keyframes`
  0%, 20%, 50%, 80%, 100% {
    transform: translateY(0);
  }
  40% {
    transform: translateY(-30px);
  }
  60% {
    transform: translateY(-15px);
  }
`;

// A page to start a new session for a game

const Game: React.FC = () => {
    const { gameId } = useParams<{ gameId: string }>();
    const dispatch = useAppDispatch();
    const navigate = useNavigate();
    const [playerName, setPlayerName] = useState('');
    const [duration, setDuration] = useState(60);

    const handleStartSession = async () => {
        const response = await dispatch(beginSession({ gameId: Number(gameId), playerName, duration }));
        const sessionId = response.payload.sessionId;
        navigate(`/sessions/${sessionId}`);
    };

    return (
        <Box
            sx={{
                display: 'flex',
                flexDirection: 'column',
                alignItems: 'center',
                justifyContent: 'center',
                height: '80vh',
                textAlign: 'center',
            }}
        >
            <SportsEsportsIcon sx={{ fontSize: 80, color: '#FE6B8B', animation: `${bounce} 2s infinite` }} />
            <Typography
                variant="h4"
                gutterBottom
                sx={{
                    mt: 2,
                    background: 'linear-gradient(45deg, #FE6B8B 30%, #FF8E53 90%)',
                    WebkitBackgroundClip: 'text',
                    WebkitTextFillColor: 'transparent',
                    fontFamily: 'Roboto, sans-serif',
                    fontWeight: 'bold',
                }}
            >
                Start a Session
            </Typography>
            <Grid container spacing={1} alignItems="center" justifyContent="center" sx={{ maxWidth: 600 }}>
                <Grid item xs={12}>
                    <Box sx={{ display: 'flex', alignItems: 'center' }}>
                        <PersonIcon sx={{ mr: 1, color: '#FF8E53' }} />
                        <TextField
                            label="Player Name"
                            value={playerName}
                            onChange={(e) => setPlayerName(e.target.value)}
                            fullWidth
                            variant="outlined"
                            margin="normal"
                        />
                    </Box>
                </Grid>
                <Grid item xs={12}>
                    <Box sx={{ display: 'flex', alignItems: 'center' }}>
                        <TimerIcon sx={{ mr: 1, color: '#FF8E53' }} />
                        <TextField
                            label="Duration (seconds)"
                            type="number"
                            value={duration}
                            onChange={(e) => setDuration(Number(e.target.value))}
                            fullWidth
                            variant="outlined"
                            margin="normal"
                            inputProps={{ min: 0 }}
                        />
                    </Box>
                </Grid>
                <Grid item xs={12}>
                    <Button
                        onClick={handleStartSession}
                        fullWidth
                        sx={{
                            mt: 2,
                            position: 'relative',
                            overflow: 'hidden',
                            color: 'red',
                            border: '2px solid transparent',
                            borderRadius: '8px',
                            fontWeight: 'bold',
                            background: '#F5F5F5',
                            cursor: 'pointer',
                            '&::before': {
                                content: '""',
                                position: 'absolute',
                                top: 0,
                                left: '-100%',
                                width: '100%',
                                height: '100%',
                                background: 'linear-gradient(45deg, #FF8E53 30%, #FE6B8B 90%)',
                                transition: 'left 0.8s ease-in-out',
                                zIndex: 0,
                            },
                            '&:hover::before': {
                                left: 0,
                            },
                            '&:hover': {
                                color: 'transparent',
                                border: '2px solid transparent',
                            },
                            '& > *': {
                                position: 'relative',
                                zIndex: 1,
                            },
                        }}
                    >
                        Start Session
                    </Button>
                </Grid>
            </Grid>
        </Box>
    );
};

export default Game;