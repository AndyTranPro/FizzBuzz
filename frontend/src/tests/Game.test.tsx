import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter, useNavigate, useParams } from 'react-router-dom';
import Game from '../pages/Game';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(),
    useParams: jest.fn(),
}));

jest.mock('../features/sessionSlice', () => ({
    beginSession: jest.fn(),
}));

const mockStore = configureStore([]);
const mockNavigate = useNavigate as jest.Mock;
const mockUseParams = useParams as jest.Mock;

const renderComponent = (store: any, route: string) => {
    return render(
        <Provider store={store}>
            <MemoryRouter initialEntries={[route]}>
                <Game />
            </MemoryRouter>
        </Provider>
    );
};

describe('Game Component', () => {
    let store: any;

    beforeEach(() => {
        store = mockStore({});
        store.dispatch = jest.fn();
        mockNavigate.mockReset();
        mockUseParams.mockReturnValue({ gameId: '1' });
    });

    test('renders correctly and matches snapshot', () => {
        const { asFragment } = renderComponent(store, '/games/1');
        expect(asFragment()).toMatchSnapshot();
    });

    test('renders player name and duration input fields', () => {
        renderComponent(store, '/games/1');

        expect(screen.getByLabelText(/Player Name/i)).toBeInTheDocument();
        expect(screen.getByLabelText(/Duration \(seconds\)/i)).toBeInTheDocument();
    });

    test('updates player name and duration when inputs change', () => {
        renderComponent(store, '/games/1');

        const playerNameInput = screen.getByLabelText(/Player Name/i);
        const durationInput = screen.getByLabelText(/Duration \(seconds\)/i);

        fireEvent.change(playerNameInput, { target: { value: 'John Doe' } });
        fireEvent.change(durationInput, { target: { value: 120 } });

        expect(playerNameInput).toHaveValue('John Doe');
        expect(durationInput).toHaveValue(120);
    });
});
