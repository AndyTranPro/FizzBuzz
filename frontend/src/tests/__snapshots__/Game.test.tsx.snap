// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1xei2fu"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1fmtb0r-MuiSvgIcon-root"
      data-testid="SportsEsportsIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m21.58 16.09-1.09-7.66C20.21 6.46 18.52 5 16.53 5H7.47C5.48 5 3.79 6.46 3.51 8.43l-1.09 7.66C2.2 17.63 3.39 19 4.94 19c.68 0 1.32-.27 1.8-.75L9 16h6l2.25 2.25c.48.48 1.13.75 1.8.75 1.56 0 2.75-1.37 2.53-2.91M11 11H9v2H8v-2H6v-1h2V8h1v2h2zm4-1c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1m2 3c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1"
      />
    </svg>
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1b175wn-MuiTypography-root"
    >
      Start a Session
    </h4>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1h6ab69-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-70qvj9"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-4kr8b6-MuiSvgIcon-root"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
            />
          </svg>
          <div
            class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r0:"
              id=":r0:-label"
            >
              Player Name
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Player Name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-70qvj9"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-4kr8b6-MuiSvgIcon-root"
            data-testid="TimerIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M9 1h6v2H9zm10.03 6.39 1.42-1.42c-.43-.51-.9-.99-1.41-1.41l-1.42 1.42C16.07 4.74 14.12 4 12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9 9-4.03 9-9c0-2.12-.74-4.07-1.97-5.61M13 14h-2V8h2z"
            />
          </svg>
          <div
            class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r1:"
              id=":r1:-label"
            >
              Duration (seconds)
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                min="0"
                type="number"
                value="60"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Duration (seconds)
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-n26n11-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Start Session
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
